// @mui/x-charts/context/CartesianProvider/CartesianProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CartesianContextProvider = CartesianContextProvider;
var React = _interopRequireWildcard(require("react"));
var _useDrawingArea = require("../../hooks/useDrawingArea");
var _useSeries = require("../../hooks/useSeries");
var _CartesianContext = require("./CartesianContext");
var _computeValue = require("./computeValue");
var _useXExtremumGetter = require("../PluginProvider/useXExtremumGetter");
var _PluginProvider = require("../PluginProvider");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function CartesianContextProvider(props) {
  const {
    xAxis,
    yAxis,
    dataset,
    children
  } = props;
  const formattedSeries = (0, _useSeries.useSeries)();
  const drawingArea = (0, _useDrawingArea.useDrawingArea)();
  const xExtremumGetters = (0, _useXExtremumGetter.useXExtremumGetter)();
  const yExtremumGetters = (0, _PluginProvider.useYExtremumGetter)();
  const xValues = React.useMemo(() => (0, _computeValue.computeValue)({
    drawingArea,
    formattedSeries,
    axis: xAxis,
    extremumGetters: xExtremumGetters,
    dataset,
    axisDirection: 'x'
  }), [drawingArea, formattedSeries, xAxis, xExtremumGetters, dataset]);
  const yValues = React.useMemo(() => (0, _computeValue.computeValue)({
    drawingArea,
    formattedSeries,
    axis: yAxis,
    extremumGetters: yExtremumGetters,
    dataset,
    axisDirection: 'y'
  }), [drawingArea, formattedSeries, yAxis, yExtremumGetters, dataset]);
  const value = React.useMemo(() => ({
    isInitialized: true,
    data: {
      xAxis: xValues.axis,
      yAxis: yValues.axis,
      xAxisIds: xValues.axisIds,
      yAxisIds: yValues.axisIds
    }
  }), [xValues, yValues]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_CartesianContext.CartesianContext.Provider, {
    value: value,
    children: children
  });
}
